docs: Complete rewrite of PROXY protocol guide

polarathene opened the pull request:
# Description

Based on the rather long discussion at https://github.com/docker-mailserver/docker-mailserver/issues/3866

The existing docs are a bit dated:
- Postfix has support for PROXY protocol v2 now (Since Postfix 3.5).
- Incorrect information at the end regarding an IMAPS port conflicting with SMTP related services `amavis` / `postscreen`.
- Dovecot override config was a bit lacking (_earlier services, no need for timeout, trusted network setting using invalid syntax `,`_)

**Added:**
- More technical insights and caveats for those it'd benefit.
- Includes a security section about forgery risk and monitoring gotchas (Roundcube + Fail2Ban) with advice to resolve.
- Small section on verifying working configuration.
- Dovecot + Postfix config for enabling PROXY protocol support + warning and alternative config advice for when separate ports may be needed (_someone may contribute that as a feature to DMS v15+_)

## Type of change

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] **I have added information about changes made in this PR to `CHANGELOG.md`**

Can bundle in changelog entry referencing various docs PRs 👍

polarathene commented:
Direct links for comparison:
- https://docker-mailserver.github.io/docker-mailserver/v13.3/examples/tutorials/mailserver-behind-proxy/
- https://pullrequest-3882--dms-doc-previews.netlify.app/examples/tutorials/mailserver-behind-proxy/

cfis commented on docs/content/examples/tutorials/mailserver-behind-proxy.md:
```diff
@@ -2,126 +2,410 @@
+    When working with Layer 7 and a protocol like HTTP, it is possible to inspect a protocol header like [`Forwarded`][networking::http-header::forwarded] (_or it's predecessor: [`X-Forwarded-For`][networking::http-header::x-forwarded-for]_). At a lower level with Layer 4, that information is not available and we are routing traffic agnostic to the application protocol being proxied.
+
+    A proxy can prepend the [PROXY protocol][networking::spec:proxy-protocol] header to the TCP/UDP connection as it is routed to the service, which must be configured to be compatible with PROXY protocol (_often this adds a restriction that connections must provide the header, otherwise they're rejected_).
+
```
I would remove "(_often this adds a restriction that connections must provide the header, otherwise they're rejected_)." I don't think its often, I think its always. But that is already captured by the sentence above.

polarathene replied:
> I don't think its often, I think its always.

What source do you have to say always? My statement is valid, here is why it varies:

- [Traefik entrypoints support the PROXY header as optional](https://doc.traefik.io/traefik/routing/entrypoints/#proxyprotocol). I'm able to enable it and connect without PROXY protocol in use from curl.
- Whereas the [equivalent for Caddy](https://caddyserver.com/docs/json/apps/http/servers/listener_wrappers/proxy_protocol/) will allow normal connections without PROXY protocol, unless they're in the `allow` list in which case they're expected to only use PROXY protocol.
  - I can allow Traefik to connect with PROXY protocol, and in a separate container route through Traefik or alternatively connect to Caddy directly without PROXY protocol. So long as I didn't allow PROXY protocol for the curl container by configuring the entire subnet, that direct connection via curl works fine.
- Postfix and Dovecot however only accept connections using PROXY protocol for ports that enable it. They're not as flexible.

cfis replied:
I base my comment on this paragraph from the spec (https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt):

> The receiver MUST be configured to only receive the protocol described in this
> specification and MUST not try to guess whether the protocol header is present
> or not. This means that the protocol explicitly prevents port sharing between
> public and private access. Otherwise it would open a major security breach by
> allowing untrusted parties to spoof their connection addresses. The receiver
> SHOULD ensure proper access filtering so that only trusted proxies are allowed
> to use this protocol.

polarathene replied:
> This means that the protocol explicitly prevents port sharing between public and private access.
> Otherwise it would open a major security breach by allowing untrusted parties to spoof their connection addresses.
> The receiver SHOULD ensure proper access filtering so that only trusted proxies are allowed to use this protocol.

The `SHOULD` making it advised rather than mandatory resolves the issue though?

If the receiver only allows a connection from a trusted host, you could require the PROXY protocol header only on those connections. But what is the risk in making the header optional when you're trusting that host regardless?

Likewise, if you reject a connection from a host providing a PROXY header when they're not trusted, what is the risk in allowing them to connect without the PROXY header present?

---

The upcoming Caddy 2.8 release changed their PROXY protocol dependency to the same that Traefik is using. Thus they'll also have the capability for ports to accept an optional PROXY header that's accepted when the remote host is trusted by the reverse proxy software.

The Go library used is [`go-proxyproto`](https://github.com/pires/go-proxyproto), and this flexibility is offered via the [Policy](https://github.com/pires/go-proxyproto/blob/v0.7.0/policy.go#L18) feature (_landed [roughly a year ago](https://github.com/pires/go-proxyproto/issues/95)_). You can see the [implementation for Caddy here](https://github.com/caddyserver/caddy/pull/5915/files).

There wasn't much discussion about security concerns or the spec you're quoting, but there doesn't seem to be any major concerns with that flexibility via policy since it still implements the `SHOULD` to ensure PROXY protocol is only valid with trusted connections.

---

Regardless, the spec doesn't reflect actual popular software support in the wild which our docs are noting that the restriction is not always imposed.

So agree to leave as-is?

cfis replied:
I think the first sentence is pretty clear:

> The receiver MUST be configured to only receive the protocol described in this
> specification and MUST not try to guess whether the protocol header is present
> or not.

I read the SHOULD as a secondary concern, it does not impact that first sentence in anyway. Thus, I understand why Postfix and Dovecot have implemented proxy support the way they have.

I think the Caddy code you posted is clearly in violation of the spec (not that it matters what I think of course). I would guess it was implemented by someone who didn't read the spec and likely didn't consider the security implications.

polarathene replied:
> I would guess it was implemented by someone who didn't read the spec and likely didn't consider the security implications.

What security implications?

If you only follow the MUST requirements of the spec and ignore the SHOULD, you have less security 🤷‍♂️ 

I see no security concerns with the policy approach that `go-proxyproto` implements.

---

> Either leave as is, or maybe add text "according to the spec the port should always ... but some implementation allow ... "

Leaving as-is 👍 

The docs are focused on PROXY protocol compatibility with different reverse proxies / software relevant to DMS. I don't see any value in the expectations of the spec when they're not well justified (_specifications can have flaws or become outdated too_).

I have [raised an issue at HAProxy](https://github.com/haproxy/haproxy/issues/2450), perhaps they'll update it accordingly or weigh in with something that justifies the current wording.

cfis replied:
The spec has been updated 11 times in ten years...guessing its not a flaw in the spec.  Excellent that you raised an issue, hope it gets a response.

cfis commented on docs/content/examples/tutorials/mailserver-behind-proxy.md:
```diff
@@ -2,126 +2,410 @@
+    #!/bin/bash
+
+    # Duplicate the config for the submission(s) service ports (587 / 465) with adjustments for the PROXY ports (10587 / 10465) and `syslog_name` setting:
+    postconf -Mf submission/inet | sed -e s/^submission/10587/ -e 's/submission/submission-proxyprotocol/' >> /etc/postfix/master.cf
```
I assume you run this inside the container? Wouldn't it be lost when the container is restarted?

polarathene replied:
> I assume you run this inside the container?

Yes, as the docs mention prior, this is `user-patches.sh`. I've annotated the codefence and provided a link reference to the relevant docs page, so the reader should already have this context.

![image](https://github.com/docker-mailserver/docker-mailserver/assets/5098581/d719d323-cdcf-4cfd-8da8-b709ac20faad)

> Wouldn't it be lost when the container is restarted?

It would be lost when the container is recreated, which is what we advise for "restarting" (destroy the previous instance).

Otherwise it'd be appended each time since the container persists internal state regardless of a volume mount when it's restarted (_or a more manual stop/start_). Which is pretty bad, we have several issues open about such bugs.

Ideally we'd have our internal base config stored elsewhere to be copied to the proper locations at each restart so that we can avoid that issue. It'll be a breaking change though as some users volume mount configs directly to the internal location and some of those are files we may modify, addressing the issue may wipe their direct mounted volume content (_one would hope they have a backup, or at least read breaking changes in our changelog_).

cfis replied:
Oh I see, missed it was part of user-patches.sh

cfis commented:
Added  a few comments, but overall looks godo to me!

polarathene commented:
> Added a few comments, but overall looks good to me! Thanks for writing this up.

Thanks for the feedback! 😁

I know it's quite a bit to go over, so I really appreciate it! Writing docs is an exhausting process for me to tackle 😩 

I think my replies have addressed the feedback you gave, but if you feel I could improve on the content or disagree in anyway let me know :)

polarathene commented:
There is an AWS specific gotcha with their load balancer with PROXY protocol that affects SMTP, not sure if it's relevant for our docs though. I've not got the time to verify if it's still accurate.

https://github.com/pires/go-proxyproto#special-notes

It was added in June 2020: https://github.com/pires/go-proxyproto/pull/25
